### PR TITLE
Replace all `Optional` with `Union`

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

Replace all `Optional` with `Union`

We have both in this project: `Optional[int]` on `port` and `Union[int, None]` on `explict_port` is not so great.  This is a short lived change since `pyupgrade` is going to change them all to `type | None` once we remove support for older python.

## Are there changes in behavior for the user?

no

## Related issue number

https://github.com/aio-libs/yarl/pull/1084#discussion_r1743316773
https://github.com/aio-libs/yarl/pull/1112#discussion_r1746859083

### CODE REVIEW

**Comment by bdraco** [comment]
Not sure what kind of changelog message this should get.

Also near midnight here so I'll have to pick this up tomorrow.

**Comment by webknjaz** [comment]
> Not sure what kind of changelog message this should get.

I'd say “contrib” since it's improving the readability for contributors w/o any influence on how the type checkers would perceive this.

File: yarl/_url.py
Comment (webknjaz) [inline yarl/_url.py:1]: Time to stop importing `Optional`?

File: CHANGES/1095.contrib.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Replace all :type:`Optional` with :type:`Union` -- by :user:`bdraco`.
```
Comment (webknjaz) [inline CHANGES/1095.contrib.rst:1]:
```suggestion
Replaced all :type:`Optional` with :type:`Union` -- by :user:`bdraco`.
```

Oh, I didn't realize you can link types like that..

**Review by webknjaz — APPROVED** [review]

File: CHANGES/1095.contrib.rst (hunk 1 above)
Comment (bdraco) [reply CHANGES/1095.contrib.rst:1]: It was an attempt but actually I don't think you can.

```
/home/docs/checkouts/readthedocs.org/user_builds/yarl/checkouts/1095/docs/changes.rst:10: WARNING: py:type reference target not found: Optional [ref.type]
/home/docs/checkouts/readthedocs.org/user_builds/yarl/checkouts/1095/docs/changes.rst:10: WARNING: py:type reference target not found: Union [ref.type]
```

File: CHANGES/1095.contrib.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Replaced all :type:`Optional` with :type:`Union` -- by :user:`bdraco`.
```
Comment (bdraco) [inline CHANGES/1095.contrib.rst:1]:
```suggestion
Replaced all ``Optional`` with ``Union`` -- by :user:`bdraco`.
```

was worth a shot but didn't actually work

File: CHANGES/1095.contrib.rst (hunk 1 above)
Comment (webknjaz) [reply CHANGES/1095.contrib.rst]: Oh, okay. That's what I assumed :)

You'd need to declare the corresponding role. You need to go for 
```rst
:class:`~typing.Optional`
```

In general, the best way to know if an object exists is to call intersphinx from command line or view as html on my helper static site..

Comment (webknjaz) [reply CHANGES/1095.contrib.rst]: @bdraco missed this comment with the working syntax?

Comment (bdraco) [reply CHANGES/1095.contrib.rst]: Yup. Just saw it now

Comment (bdraco) [reply CHANGES/1095.contrib.rst]: https://github.com/aio-libs/yarl/actions/runs/10746477913 is still going.  If it gets past the wheel combine this time, I'll cancel it and rework the changelog message and start the release process over

Comment (bdraco) [reply CHANGES/1095.contrib.rst]: https://github.com/aio-libs/yarl/pull/1121

**Pull request merged**